File: banchus-app/src/main/java/pe/nanamochi/banchus/service/BeatmapService.java
package pe.nanamochi.banchus.service;

import java.nio.file.Path;
import java.time.Instant;
import java.util.List;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import pe.nanamochi.banchus.database.entity.Beatmap;
import pe.nanamochi.banchus.database.entity.Beatmapset;
import pe.nanamochi.banchus.database.repository.BeatmapRepository;
import pe.nanamochi.banchus.domain.dto.OsuApiBeatmap;
import pe.nanamochi.banchus.mapper.BeatmapMapper;
import pe.nanamochi.banchus.service.infra.OsuApiClient;
import pe.nanamochi.banchus.util.Security;

@Slf4j
@Service
@RequiredArgsConstructor
public class BeatmapService {

  private final BeatmapRepository beatmapRepository;
  private final BeatmapMapper beatmapMapper;
  private final StorageService storageService;
  private final OsuApiClient osuApiClient;
  private final BeatmapsetService beatmapsetService;

  public Beatmap save(Beatmap beatmap) {
    return beatmapRepository.save(beatmap);
  }

  public Beatmap save(OsuApiBeatmap apiBeatmap, Beatmapset beatmapset) {
    Beatmap beatmap = beatmapMapper.fromApi(apiBeatmap);
    beatmap.setBeatmapset(beatmapset);
    return save(beatmap);
  }

  public Beatmap update(Beatmap beatmap) {
    if (!beatmapRepository.existsById(beatmap.getId())) {
      throw new IllegalArgumentException("Beatmap not found: " + beatmap.getId());
    }
    return save(beatmap);
  }

  public Optional<Beatmap> findById(int id) {
    return beatmapRepository.findById(id);
  }

  public Optional<Beatmap> findByMd5(String md5) {
    return beatmapRepository.findByMd5(md5);
  }

  public Optional<byte[]> getOrDownloadOsuFile(int beatmapId, String expectedMd5) {
    Optional<byte[]> localFile = storageService.getBeatmap(beatmapId);

    if (localFile.isPresent()) {
      byte[] data = localFile.get();
      if (expectedMd5 == null || Security.getMd5FromBytes(data).equalsIgnoreCase(expectedMd5)) {
        return localFile;
      }
      log.debug("Local .osu file for {} does not match MD5. Redownloading...", beatmapId);
    }

    return osuApiClient
        .getOsuFile(beatmapId)
        .map(
            downloaded -> {
              storageService.uploadBeatmap(beatmapId, downloaded);
              return downloaded;
            });
  }

  public Path getBeatmapPath(int beatmapId) {
    return storageService.getBeatmapPath(beatmapId);
  }

  public Beatmap getOrCreateBeatmap(String beatmapMd5) {
    Beatmap beatmap =
        findByMd5(beatmapMd5)
            .orElseGet(
                () -> {
                  log.debug("Beatmap {} not found. Querying osu!api...", beatmapMd5);

                  OsuApiBeatmap apiData = osuApiClient.getBeatmap(beatmapMd5).orElse(null);
                  if (apiData == null) return null;

                  Beatmapset beatmapset =
                      beatmapsetService
                          .findById(apiData.getBeatmapsetId())
                          .orElseGet(() -> beatmapsetService.save(apiData));

                  List<OsuApiBeatmap> setBeatmaps =
                      osuApiClient.getBeatmaps(apiData.getBeatmapsetId());

                  Beatmap target = null;
                  for (OsuApiBeatmap b : setBeatmaps) {
                    Beatmap current =
                        findByMd5(b.getFileMd5())
                            .orElseGet(
                                () -> {
                                  Beatmap newMap = save(b, beatmapset);
                                  getOrDownloadOsuFile(b.getBeatmapId(), b.getFileMd5());
                                  return newMap;
                                });

                    if (b.getFileMd5().equals(beatmapMd5)) target = current;
                  }
                  return target;
                });

    if (beatmap != null) {
      beatmap = updateBeatmapIfOutdated(beatmap);
    }

    return beatmap;
  }

  private Beatmap updateBeatmapIfOutdated(Beatmap beatmap) {
    List<OsuApiBeatmap> osuApiBeatmaps = osuApiClient.getBeatmaps(beatmap.getBeatmapset().getId());
    if (osuApiBeatmaps.isEmpty()) return beatmap;

    Instant localLastUpdate = beatmap.getBeatmapset().getLastUpdated();
    Instant remoteLastUpdate =
        osuApiBeatmaps.stream()
            .map(OsuApiBeatmap::getLastUpdate)
            .max(Instant::compareTo)
            .orElse(localLastUpdate);

    if (localLastUpdate.isBefore(remoteLastUpdate)) {
      log.debug("Updating beatmapset {} (Outdated)", beatmap.getBeatmapset().getId());

      Beatmapset beatmapset = beatmap.getBeatmapset();
      beatmapset.setLastUpdated(remoteLastUpdate);
      beatmapsetService.save(beatmapset);

      for (OsuApiBeatmap b : osuApiBeatmaps) {
        getOrDownloadOsuFile(b.getBeatmapId(), b.getFileMd5());

        Beatmap local = findByMd5(b.getFileMd5()).orElse(null);
        if (local != null) {
          local.setLastUpdated(b.getLastUpdate());
          local.setStarRating(b.getDifficultyRating());
          update(local);

          if (b.getFileMd5().equals(beatmap.getMd5())) {
            beatmap = local;
          }
        }
      }
    }

    return beatmap;
  }
}

File: banchus-app/src/main/kotlin/pe/nanamochi/banchus/service/BeatmapService.kt
package pe.nanamochi.banchus.service

import com.github.michaelbull.result.Err
import com.github.michaelbull.result.Result
import com.github.michaelbull.result.binding
import com.github.michaelbull.result.getOrElse
import com.github.michaelbull.result.map
import com.github.michaelbull.result.onSuccess
import com.github.michaelbull.result.orElse
import com.github.michaelbull.result.toResultOr
import jakarta.transaction.Transactional
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import pe.nanamochi.banchus.database.entity.Beatmap
import pe.nanamochi.banchus.database.entity.Beatmapset
import pe.nanamochi.banchus.database.repository.BeatmapRepository
import pe.nanamochi.banchus.domain.errors.BeatmapNotFound
import pe.nanamochi.banchus.domain.errors.DomainMessage
import pe.nanamochi.banchus.domain.errors.InternalError
import pe.nanamochi.banchus.dto.external.OsuApiBeatmap
import pe.nanamochi.banchus.infrastructure.clients.OsuApiClient
import pe.nanamochi.banchus.mapper.BeatmapMapper
import pe.nanamochi.banchus.util.runDatabaseCatching
import pe.nanamochi.banchus.util.toMd5

@Service
class BeatmapService(
    private val beatmapRepository: BeatmapRepository,
    private val beatmapsetService: BeatmapsetService,
    private val beatmapMapper: BeatmapMapper,
    private val storageService: StorageService,
    private val osuApiClient: OsuApiClient,
) {
    private val log = LoggerFactory.getLogger(javaClass)

    fun findById(id: Int): Result<Beatmap, BeatmapNotFound> =
        beatmapRepository.findBeatmapById(id).toResultOr { BeatmapNotFound }

    fun findByMd5(md5: String): Result<Beatmap, BeatmapNotFound> =
        beatmapRepository.findByMd5(md5).toResultOr { BeatmapNotFound }

    fun create(beatmap: Beatmap): Result<Beatmap, DomainMessage> = runDatabaseCatching {
        beatmapRepository.save(beatmap)
    }

    fun create(apiBeatmap: OsuApiBeatmap, beatmapset: Beatmapset): Result<Beatmap, DomainMessage> {
        val beatmap = beatmapMapper.fromApi(apiBeatmap).apply { this.beatmapset = beatmapset }
        return create(beatmap)
    }

    fun update(beatmap: Beatmap): Result<Beatmap, DomainMessage> =
        if (beatmapRepository.existsById(beatmap.id)) {
            runDatabaseCatching { beatmapRepository.save(beatmap) }
        } else {
            Err(BeatmapNotFound)
        }

    @Transactional
    fun incrementStats(id: Int, passed: Boolean): Result<Unit, DomainMessage> =
        runDatabaseCatching { beatmapRepository.incrementStats(id, passed) }.map {}

    fun getOrCreateBeatmap(beatmapMd5: String): Result<Beatmap, DomainMessage> = binding {
        val localBeatmap =
            findByMd5(beatmapMd5).getOrElse { _ ->
                log.debug("Beatmap {} not found local. Querying osu!api...", beatmapMd5)

                val apiData =
                    osuApiClient.getBeatmap(beatmapMd5).toResultOr { BeatmapNotFound }.bind()

                val setId = apiData.beatmapsetId.toResultOr { BeatmapNotFound }.bind()

                val beatmapset =
                    beatmapsetService
                        .findById(setId)
                        .orElse { beatmapsetService.create(apiData) }
                        .bind()

                osuApiClient
                    .getBeatmaps(setId)
                    .map { apiMap ->
                        val currentMd5 = apiMap.fileMd5.toResultOr { InternalError }.bind()

                        val current =
                            findByMd5(currentMd5).getOrElse { _ ->
                                val newMap = create(apiMap, beatmapset).bind()
                                val bId = apiMap.beatmapId.toResultOr { InternalError }.bind()
                                getOrDownloadOsuFile(bId, currentMd5).bind()
                                newMap
                            }
                        current
                    }
                    .find { it.md5.equals(beatmapMd5, ignoreCase = true) }
                    ?: Err(BeatmapNotFound).bind()
            }

        log.debug("Found local beatmap for MD5 {}: {}", beatmapMd5, localBeatmap.id)

        updateBeatmapIfOutdated(localBeatmap).bind()
    }

    fun getOrDownloadOsuFile(
        beatmapId: Int,
        expectedMd5: String?,
    ): Result<ByteArray, DomainMessage> = binding {
        storageService.getBeatmap(beatmapId).onSuccess { data ->
            val isValid =
                expectedMd5?.let { md5 -> data.toMd5().equals(md5, ignoreCase = true) } ?: true

            if (isValid) return@binding data
            log.debug("Local .osu file for {} does not match MD5. Redownloading...", beatmapId)
        }

        val downloaded = osuApiClient.getOsuFile(beatmapId).toResultOr { BeatmapNotFound }.bind()

        storageService.saveBeatmap(beatmapId, downloaded).bind()

        downloaded
    }

    private fun updateBeatmapIfOutdated(beatmap: Beatmap): Result<Beatmap, DomainMessage> =
        binding {
            val beatmapset = beatmap.beatmapset ?: return@binding beatmap

            val osuApiBeatmaps = osuApiClient.getBeatmaps(beatmapset.id)
            if (osuApiBeatmaps.isEmpty()) return@binding beatmap

            val remoteLastUpdate =
                osuApiBeatmaps.mapNotNull { it.lastUpdate }.maxOrNull() ?: beatmapset.lastUpdated

            if (beatmapset.lastUpdated.isBefore(remoteLastUpdate)) {
                log.debug("Updating beatmapset {} (Outdated)", beatmapset.id)

                beatmapset.lastUpdated = remoteLastUpdate
                beatmapsetService.update(beatmapset).bind()

                var updatedTarget = beatmap

                for (apiMap in osuApiBeatmaps) {
                    val apiMd5 = apiMap.fileMd5 ?: continue
                    val apiId = apiMap.beatmapId ?: continue

                    getOrDownloadOsuFile(apiId, apiMd5)

                    findByMd5(apiMd5).onSuccess { local ->
                        local.apply {
                            lastUpdated = apiMap.lastUpdate ?: remoteLastUpdate
                            starRating = apiMap.difficultyRating ?: starRating
                        }
                        update(local).bind()

                        if (apiMd5.equals(beatmap.md5, ignoreCase = true)) {
                            updatedTarget = local
                        }
                    }
                }
                return@binding updatedTarget
            }

            beatmap
        }
}
